gateway_tests/process/test_logging.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import annotations
+
+import contextlib
+import dataclasses
+import datetime
+import logging
+import socket
+import tempfile
+import threading
+from typing import Any, ContextManager, Optional
+
+import pytest
+
+from .. import conftest
+
+logger = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def listen_on_port(port, encoding="latin-1") -> ContextManager[list[str]]:
+    """Listen on TCP port `port` for caPutLog data."""
+    data = []
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind(('127.0.0.1', port))
+
+    def listen():
+        sock.listen(1)
+        client, addr = sock.accept()
+        logger.warning("TCP accepted client %s", addr)
+        while True:
+            read = client.recv(4096)
+            logger.info("caPutLog TCP server received %s", read)
+            if not data:
+                client.close()
+                return
+            data.append(read.decode(encoding))
+
+    threading.Thread(target=listen, daemon=True).start()
+    yield data
+    sock.close()
+
+
+@dataclasses.dataclass
+class Caput:
+    """A single caputlog line, parsed into its parts."""
+    date: datetime.datetime
+    user: str
+    host: str
+    pvname: str
+    value: str
+    old: Optional[str] = None
+
+    @classmethod
+    def from_line(cls, line: str) -> Caput:
```

Comment:
I like this pattern of making a self-parsing dataclass

Reply:
I'm pretty fond of it, too! 👍